File: nix/nixpkgs/devlogs-lib/devlogs.lua
local M = {}

local window = ""
local pane = vim.env.TMUX_PANE
if pane then
	local out = vim.fn.system({ "tmux", "display-message", "-t", pane, "-p", "#{window_index}" })
	if vim.v.shell_error == 0 then
		window = vim.trim(out)
	end
end

local function log(level, component, instance, msg)
	local tag = component .. "{" .. instance .. "}"
	if window ~= "" then
		tag = tag .. "(@" .. window .. ")"
	end
	local formatted = ("[devlogs] %s %s: %s"):format(level:upper(), tag, msg)
	-- all levels use user.info; macOS unified logging drops user.debug from history
	vim.fn.system({ "logger", "-t", "devlogs", "-p", "user.info", formatted })
end

function M.new(component)
	component = component or vim.env.DEVLOGS_COMPONENT or "unknown"
	local instance = vim.env.DEVLOGS_INSTANCE or "-"
	local logger = {}
	for _, level in ipairs({ "debug", "info", "warning", "error" }) do
		logger[level] = function(msg)
			log(level, component, instance, msg)
		end
	end
	return logger
end

return M

File: nix/nixpkgs/devlogs-lib/devlogs.ts
import { spawnSync } from "child_process";

type Level = "debug" | "info" | "warning" | "error";

const PRIORITY_MAP: Record<Level, string> = {
  debug: "user.info", // macOS unified logging drops user.debug from history
  info: "user.info",
  warning: "user.warning",
  error: "user.err",
};

let _window = "";
const pane = process.env.TMUX_PANE;
if (pane) {
  try {
    const result = spawnSync(
      "tmux",
      ["display-message", "-t", pane, "-p", "#{window_index}"],
      { encoding: "utf-8", timeout: 2000 },
    );
    if (result.status === 0 && result.stdout) {
      _window = result.stdout.trim();
    }
  } catch {}
}

function log(level: Level, component: string, instance: string, msg: string) {
  let tag = `${component}{${instance}}`;
  if (_window) {
    tag = `${component}{${instance}}(@${_window})`;
  }
  const formatted = `[devlogs] ${level.toUpperCase()} ${tag}: ${msg}`;
  try {
    spawnSync("logger", ["-t", "devlogs", "-p", PRIORITY_MAP[level], formatted], {
      timeout: 2000,
    });
  } catch {}
}

export interface Logger {
  debug(msg: string): void;
  info(msg: string): void;
  warning(msg: string): void;
  error(msg: string): void;
}

export function createLogger(component?: string, instance?: string): Logger {
  const comp = component || process.env.DEVLOGS_COMPONENT || "unknown";
  const inst = instance || process.env.DEVLOGS_INSTANCE || "-";
  return {
    debug: (msg) => log("debug", comp, inst, msg),
    info: (msg) => log("info", comp, inst, msg),
    warning: (msg) => log("warning", comp, inst, msg),
    error: (msg) => log("error", comp, inst, msg),
  };
}
